src/main/java/org/opensearch/knn/indices/ModelDao.java

```diff
@@ -404,8 +409,12 @@ public void get(String modelId, ActionListener<GetModelResponse> actionListener)
          */
         @Override
         public void search(SearchRequest request, ActionListener<SearchResponse> actionListener) {
-            request.indices(MODEL_INDEX_NAME);
-            client.search(request, actionListener);
+            // temporary setting thread context to default, this is needed to allow actions on model system index when security plugin is
+            // enabled
+            try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
+                request.indices(MODEL_INDEX_NAME);
+                client.search(request, actionListener);
```

Comment:
Are there any implications of stashing thread context with an asynchronous call?

Reply:
I can't find any mention of such implications. My understanding is that for caller it might look differently comparing to sync call, it will restore context without waiting for that called method to complete, but for the method it's ThreadLocal will have a context that we have reset to default.

Reply:
Yes, you are correct.